File: mobile/ios/OpenU60/Features/RouterSettings/WiFi/WiFiSettingsViewModel.swift
import SwiftUI

@Observable
@MainActor
final class WiFiSettingsViewModel {
    var config: WiFiConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    // Editable fields
    var editSSID2g: String = ""
    var editSSID5g: String = ""
    var editKey2g: String = ""
    var editKey5g: String = ""
    var editChannel2g: String = "auto"
    var editChannel5g: String = "auto"
    var editTxpower2g: String = "100"
    var editTxpower5g: String = "100"
    var editEncryption2g: String = "psk2+ccmp"
    var editEncryption5g: String = "psk2+ccmp"
    var editHidden2g: Bool = false
    var editHidden5g: Bool = false
    var editWifiOnOff: Bool = true
    var editRadio2gDisabled: Bool = false
    var editRadio5gDisabled: Bool = false
    var editWifi7Enabled: Bool = false
    var editBandwidth2g: String = "auto"
    var editBandwidth5g: String = "auto"

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        // Try native /api/wifi/status first (single call, all fields)
        if let wifiData = try? await client.getJSON("/api/wifi/status"),
           wifiData["htmode_2g"] != nil {
            config = WiFiParser.parse(wifiData)
            syncEditFields()
            isLoading = false
            return
        }

        showMessage("Failed to load WiFi settings", isError: true)
        isLoading = false
    }

    func apply() async {
        isLoading = true
        let params: [String: Any] = [
            "ssid_2g": editSSID2g,
            "ssid_5g": editSSID5g,
            "key_2g": editKey2g,
            "key_5g": editKey5g,
            "channel_2g": editChannel2g,
            "channel_5g": editChannel5g,
            "txpower_2g": editTxpower2g,
            "txpower_5g": editTxpower5g,
            "encryption_2g": editEncryption2g,
            "encryption_5g": editEncryption5g,
            "hidden_2g": editHidden2g ? "1" : "0",
            "hidden_5g": editHidden5g ? "1" : "0",
            "wifi_onoff": editWifiOnOff ? "1" : "0",
            "radio2_disabled": editRadio2gDisabled ? "1" : "0",
            "radio5_disabled": editRadio5gDisabled ? "1" : "0",
            "wifi6_switch": editWifi7Enabled ? "1" : "0",
            "htmode_2g": editBandwidth2g,
            "htmode_5g": editBandwidth5g
        ]

        do {
            let _ = try await client.putJSON("/api/wifi/settings", body: params)
            showMessage("WiFi settings applied — WiFi will restart briefly", isError: false)
            updateConfigFromEdits()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func updateConfigFromEdits() {
        config = WiFiConfig(ssid2g: editSSID2g, ssid5g: editSSID5g, key2g: editKey2g, key5g: editKey5g,
                            channel2g: editChannel2g, channel5g: editChannel5g,
                            txpower2g: editTxpower2g, txpower5g: editTxpower5g,
                            encryption2g: editEncryption2g, encryption5g: editEncryption5g,
                            wifiOnOff: editWifiOnOff, hidden2g: editHidden2g, hidden5g: editHidden5g,
                            radio2gDisabled: editRadio2gDisabled, radio5gDisabled: editRadio5gDisabled,
                            wifi7Enabled: editWifi7Enabled, bandwidth2g: editBandwidth2g, bandwidth5g: editBandwidth5g)
    }

    private func syncEditFields() {
        editSSID2g = config.ssid2g
        editSSID5g = config.ssid5g
        editKey2g = config.key2g
        editKey5g = config.key5g
        editChannel2g = config.channel2g
        editChannel5g = config.channel5g
        editTxpower2g = config.txpower2g
        editTxpower5g = config.txpower5g
        editEncryption2g = config.encryption2g
        editEncryption5g = config.encryption5g
        editHidden2g = config.hidden2g
        editHidden5g = config.hidden5g
        editWifiOnOff = config.wifiOnOff
        editRadio2gDisabled = config.radio2gDisabled
        editRadio5gDisabled = config.radio5gDisabled
        editWifi7Enabled = config.wifi7Enabled
        editBandwidth2g = config.bandwidth2g
        editBandwidth5g = config.bandwidth5g
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/wifi/WiFiSettingsViewModel.kt
package com.openu60.feature.router.wifi

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.WiFiConfig
import com.openu60.core.model.WiFiParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class WiFiSettingsState(
    val config: WiFiConfig = WiFiConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class WiFiSettingsViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(WiFiSettingsState())
    val state: StateFlow<WiFiSettingsState> = _state.asStateFlow()

    fun updateConfig(config: WiFiConfig) {
        _state.value = _state.value.copy(config = config)
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/wifi/settings")
                val config = WiFiParser.parse(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun save() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val c = _state.value.config
                val params = mapOf(
                    "ssid_2g" to c.ssid2g,
                    "ssid_5g" to c.ssid5g,
                    "key_2g" to c.key2g,
                    "key_5g" to c.key5g,
                    "channel_2g" to c.channel2g,
                    "channel_5g" to c.channel5g,
                    "txpower_2g" to c.txpower2g,
                    "txpower_5g" to c.txpower5g,
                    "encryption_2g" to c.encryption2g,
                    "encryption_5g" to c.encryption5g,
                    "wifi_onoff" to if (c.wifiOnOff) "1" else "0",
                    "hidden_2g" to if (c.hidden2g) "1" else "0",
                    "hidden_5g" to if (c.hidden5g) "1" else "0",
                    "radio2_disabled" to if (c.radio2gDisabled) "1" else "0",
                    "radio5_disabled" to if (c.radio5gDisabled) "1" else "0",
                    "wifi6_switch" to if (c.wifi7Enabled) "1" else "0",
                    "htmode_2g" to c.bandwidth2g,
                    "htmode_5g" to c.bandwidth5g,
                )
                agentClient.putJSON("/api/wifi/settings", params)
                _state.value = _state.value.copy(isLoading = false, message = "WiFi settings saved", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) save() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}
